### PR TITLE
ENH: add anderson darling test for weibull #10829

### LINKED ISSUE #10829: Extend Anderson Darling to cover Weibull distribution
**Is your feature request related to a problem? Please describe.**

The current scipy.stats implementation of Anderson Darling only covers {‘norm’,’expon’,’logistic’,’gumbel’,’gumbel_l’, gumbel_r’, ‘extreme1’}.  However according to https://www.itl.nist.gov/div898/handbook/eda/section3/eda35e.htm tables of critical values are also available for the Weibull distribution.  In R , gofstat implements these tests https://www.rdocumentation.org/packages/fitdistrplus/versions/1.0-14/topics/gofstat

**Describe the solution you'd like**

Extend Anderson Darling to cover the Weibull distribution

**Describe alternatives you've considered**

**Additional context (e.g. screenshots)**

### PR DESCRIPTION
**Pull request by zverianskii**
#### Reference issue

Closes gh-10829

#### What does this implement/fix?

AD-test for 3-param weibull_min based on [ Estimation and Tests of Fit for the Three-Parameter Weibull Distribution
Richard A. Lockhart and Michael A. Stephens ](https://www.jstor.org/stable/2346122?seq=1)

#### Additional information
R version from [fitdistrplus](https://www.rdocumentation.org/packages/fitdistrplus/versions/1.0-14/source) works only for 2-param weibull

### CODE REVIEW

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -1696,6 +1696,21 @@ def shapiro(x):
 #             Vol. 66, Issue 3, Dec. 1979, pp 591-595.
 _Avals_logistic = array([0.426, 0.563, 0.660, 0.769, 0.906, 1.010])
 
+# From Richard A. Lockhart and Michael A. Stephens "Estimation and Tests of
+#             Fit for the Three-Parameter Weibull Distribution"
+#             Journal of the Royal Statistical Society.Series B(Methodological)
+#             Vol. 56, No. 3 (1994), pp. 491-500
+_Avals_weibull = {0: array([0.292, 0.395, 0.467, 0.522, 0.617, 0.711, 0.836, 0.931]),
```
Comment (chrisb83) [inline scipy/stats/morestats.py:1703]: what are the keys of the dictionary?

File: scipy/stats/morestats.py [hunk 2]
```diff
@@ -1816,6 +1838,13 @@ def rootfunc(ab, xj, N):
         logsf = distributions.gumbel_r.logsf(w)
         sig = array([25, 10, 5, 2.5, 1])
         critical = around(_Avals_gumbel / (1.0 + 0.2/sqrt(N)), 3)
+    elif dist == 'weibull_min':
+        c, loc, scale = distributions.weibull_min.fit(y, loc=0.0)
```
Comment (chrisb83) [inline scipy/stats/morestats.py:1842]: why do you use 0 as a guess for the mean?

File: scipy/stats/tests/test_morestats.py [hunk 3]
```diff
@@ -277,6 +277,16 @@ def test_gumbel_r(self):
         assert_array_less(A1, crit1[-2:])
         assert_(A2 > crit2[-1])
 
+    def test_weibull_min(self):
+        x1 = np.array([225, 171, 198, 189, 135, 162, 135, 117, 162])
```
Comment (chrisb83) [inline scipy/stats/tests/test_morestats.py:281]: where does this testcase come from?

File: scipy/stats/morestats.py [hunk 4]
```diff
@@ -1816,6 +1838,13 @@ def rootfunc(ab, xj, N):
...
+        logcdf = distributions.weibull_min.logcdf(x=y, c=c, loc=loc, scale=scale)
+        logsf = distributions.weibull_min.logsf(x=y, c=c, loc=loc, scale=scale)
+        shape_for_critical = min(np.round(np.divide(1.0, c), 2) * 100, 50)
```
Comment (chrisb83) [inline scipy/stats/morestats.py:1845]: wouldn't this give a key error unless you hit one of the keys by chance?

**Review by chrisb83 — COMMENTED** [review]
Hi,

thanks for the PR. the paper indeed presents a method to extend the test to weibull and it would be a nice additon. I left some comments.

File: scipy/stats/morestats.py (hunk 2 above)
Comment (zveryansky) [reply scipy/stats/morestats.py:1842]: https://github.com/scipy/scipy/issues/11806 because it is the way to make MLE work for 3 param weibull

File: scipy/stats/tests/test_morestats.py (hunk 3 above)
Comment (zveryansky) [reply scipy/stats/tests/test_morestats.py:281]: both tests are from article where this method was introduced

File: scipy/stats/morestats.py (hunk 1 above)
Comment (zveryansky) [reply scipy/stats/morestats.py:1703]: it is a 100*C-values from table 1

File: scipy/stats/morestats.py (hunk 4 above)
Comment (zveryansky) [reply scipy/stats/morestats.py:1845]: c in this context is m in the article. 

it is the way to map it's values to the table ones. But you are right I forget to convert it to an integer value as in _Avals_weibull keys. will fix that

File: scipy/stats/morestats.py (hunk 1 above)
Comment (chrisb83) [reply scipy/stats/morestats.py:1703]: thanks, please add a comment when you update the code

File: scipy/stats/morestats.py (hunk 4 above)
Comment (chrisb83) [reply scipy/stats/morestats.py:1845]: isn't there some kind of interpolation needed, e.g. if `100*c = 27`?

File: scipy/stats/tests/test_morestats.py (hunk 3 above)
Comment (chrisb83) [reply scipy/stats/tests/test_morestats.py:281]: please add a reference, too. and can the test be improved to check that A1 is approximately 0.26, c = 0.42 ...

**Comment by mdhaber** [comment]
Hi @lispsil, do you plan to finish this up?

**Comment by lispsil** [comment]
@mdhaber I will have time in the end of December.

Feel free to proceed, if any questions I'll do my best with quick replies🙂

File: scipy/stats/tests/test_morestats.py (hunk 3 above)
Comment (lispsil) [reply scipy/stats/tests/test_morestats.py:288]: With current quality of estimation by MLE we can't estimate that c = 0.42

File: scipy/stats/morestats.py (hunk 4 above)
Comment (lispsil) [reply scipy/stats/morestats.py]: your are right, fixed that

File: scipy/stats/morestats.py (hunk 1 above)
Comment (chrisb83) [reply scipy/stats/morestats.py:1703]: the reference is fine, please also refer to Table 1 and the fact that the keys are 100*C. otherwise it is hard to follow

File: scipy/stats/tests/test_morestats.py (hunk 3 above)
Comment (chrisb83) [reply scipy/stats/tests/test_morestats.py:288]: hm, that does not sound very comforting. do we get at least close to A1 = 0.26 and c = 0.42? what is the output of the function?

Comment (lispsil) [reply scipy/stats/tests/test_morestats.py:288]: example 1:
paper: loc = 99.02, scale = 78.23, c = 2.38, paper_c = 0.42, A2 = 0.26 < 0.301
scipy: loc = -0.10, scale = 181.5, c = 5.98, paper_c = 0.167, A2 = 0.223 < 0.301

example 2:
paper: loc = 9.313, scale= 93.50, c = 0.763, paper_c = 1.31, A2 = 0.54 > 0.342
scipy: loc = 11.99, scale = 100.7, c = 0.75, paper_c = 1.33, A2 = 2.29 > 0.342

In the paper they introduced a bias correction for 3-param weibull estimation, as a separate feature I could code the specific .fit function for weibull_min

File: scipy/stats/morestats.py (hunk 4 above)
Comment (chrisb83) [reply scipy/stats/morestats.py]: ok, that works. (not sure if one needs `np.divide`)

File: scipy/stats/tests/test_morestats.py (hunk 3 above)
Comment (chrisb83) [reply scipy/stats/tests/test_morestats.py]: thanks for the clarification. after checking the paper, I understand that fitting the parameters is not easy and the MLE estimate is known to have problems. that is why the authors introduce a method for three cases A, B and C. so I assume the approach implemented might not be very robust. what do you think about this point?

File: scipy/stats/morestats.py (hunk 2 above)
Comment (chrisb83) [inline scipy/stats/morestats.py:1842]: I propose to write `m, loc, ...`. otherwise it is very confusing since the paper uses m and c = 1/m, especially if we write above that "Keys are c*100"

**Comment by chrisb83** [comment]
thanks for making the changes. The main open point is the parameter estimation: I have doubts that the current approach will produce stable results when I read the introduction of the paper: "... there are problems with ML estimation [...] for m unknown" and the procedure the authors describe in section 2

File: scipy/stats/tests/test_morestats.py (hunk 3 above)
Comment (mdhaber) [reply scipy/stats/tests/test_morestats.py]: gh-11806 is an example of the trouble.

> In the paper they introduced a bias correction for 3-param weibull MLE, as a separate feature I could code the specific .fit function for weibull_min

That would be really nice.

**Comment by lispsil** [comment]
@chrisb83 well..it is a matter of fact that the naive MLE is not the best way to implement .fit for 3-param weibull. Probably that's why I don't know any package that works with 3-param weibull. I am not sure it is an issue for AD test, because for me it sounds more like a general issue of what to do with 3-param weibull.

If there is no other suggestions I can start working on PR to .fit function based on this paper. @mdhaber what do you think? maybe where are better ways to implement .fit?

**Comment by mdhaber** [comment]
> maybe there are better ways to implement .fit?

You can see [here](https://github.com/scipy/scipy/issues/11782#issuecomment-729753164) that I was looking into how to fit Weibull on my own. I got to the same result as the paper does for the scale (sigma), but didn't go as deeply into the meaning of the other likelihood equations. So since SciPy's fit method is supposed to do MLE (not some other idea of "best fit"), what you found in the paper is the best way I know!

I think gh-12737 is a good example of how to add analytical formulas. It shows how different cases of user-input (e.g. `loc` is fixed, the others are free) may need to be treated separately - including falling back on the generic `super` `fit` method if no formula is available. We have (over the course of a few PRs) come up with a nice `_check_fit_input_parameters` method that should help.  It also shows how the `._fitstart` method is used to get initial guesses of parameters (if need be - you can override that, too, though, if you you need) and how you can use `_assert_less_or_close_loglike` in your tests to ensure that the analytical fit is no worse than the superclass fit. Finally, you can test the performance of your new `fit` override by adding it to the `ContinuousFitAnalyticalMLEOverride` benchmark.

**Comment by mdhaber** [comment]
Hi @lispsil were you still interested in overriding Weibull's fit method? That sounds like it would be a big improvement, as it would also address some other reported issues (gh-11806, gh-12134).

**Comment by lispsil** [comment]
@mdhaber yes, I'll try to find time in January

**Comment by lispsil** [comment]
@mdhaber 
I am working on weibull, the paper separates three situations and I am stuck with Case C
image

"the value of gamma can easily be found by iteration" any ideas on upper/lower boundaries or at least should it be integer/real number? Because in my tests it look like very important in what space we are looking for a right Gamma

**Comment by mdhaber** [comment]
Yeah that is the trick, isn't it. For some of these I've tried to find analytical upper/lower bounds, but without much success. Take a look at how `scipy.stats.rv_continuous.ppf` is implemented. Before it uses the scalar root-finder, it first iterates to find bounds. Similar in gh-13340 (search for constant multiplier `25`). If you know the function only crosses zero once in a given direction, that should work. Is that true for you? If so, could you break this out into a private function? I could use it in at least two other places (`powerlaw` and `pareto`).

**Comment by mdhaber** [comment]
@lispsil Were you still interested in pursuing this? 
If finding a bracket for scalar root finding was still a problem, please take a look at [this code](https://github.com/scipy/scipy/blob/bc9fa43b9137c4948f5091a9edcafd642455f409/scipy/stats/_continuous_distns.py#L6286-L6297) we use to find a bracket for the `scale` of the Pareto distribution. That's what I was referring to above. If there's only one sign change, it should be pretty effective at finding it.

**Comment by mdhaber** [comment]
@lispsil I've been working on finishing this up, so I'll go ahead and push the work here, if that's OK.

> "the value of gamma can easily be found by iteration" any ideas on upper/lower boundaries or at least should it be integer/real number?

Yeah, the authors really didn't give a complete plan here. I think they envisioned that readers would be fitting to data interactively rather than using automated software. Sure, by generating and investigating plots like these interactively we can determine whether these crossings exist or not, but writing code to do that reliably is difficult.

My approach is as follows:

- Improve the generic `fit` functionality (see gh-16804) by using method of moments as a guess. We could also consider using the result for optimal scale in terms of location and shape, but I haven't played around with that yet. I had some luck using other optimizers (e.g. using a gradient-based optimizer and provide the analytical gradient) but I'll save that for later.
- Use `fit`. 
- If `fit` converges to a solution with the location very close to the minimum of the data, I think we have either case B or C. If not, we have case A.
  - If case A, zero the likelihood equations using `root` for a more precise solution. We're done wit fitting
  - If case B or C, try the bias reduction procedure. If it succeeds, assume we indeed had case B, not case C.
- If fitting is successful, proceed to calculate the test statistic and complete the test.  
- If anything fails, just raise an error explaining the problem and stating that `anderson` can't continue.  It sounds like case C is rare anyway, and the authors didn't provide example data, so without looking further, I don't have a way of testing case C behavior. Also, I think aborting is the right thing to do in case C, because the authors said that the attempt to fit a Weibull distribution should be abandoned in favor of the generalized extreme value distribution. This seems to automatically reject the null hypothesis of `anderson`. I don't think they included the possibility of case C in their critical values of the test statistic because they don't indicate how the test statistic would evaluated in this case.

I can push the updates when gh-16804 gets merged, because I need those commits for this work, and I don't want to also push those as part of this PR.

**Comment by lispsil** [comment]
@mdhaber hi, sorry for the absence. Feel free to proceed, I don't expect to have time for that in the nearest future.

**Comment by mdhaber** [comment]
No problem @lispsil but would you be willing to let me know by what you think of the approach? I think I can do the code, and it would be terrific if you'd still be interested in testing it out/reviewing.

**Comment by lispsil** [comment]
@mdhaber sure:

1. looked into the new fit method; great job!
2. I don't understand your logic behind the A, B, and C. Why do we expect the location to be away from the minimum in the data? have you looked into https://www.weibull.com/hotwire/issue148/hottopics148.htm ?

**Comment by mdhaber** [comment]
@lispsil I don't understand the question:
> Why do we expect the location to be away from the minimum in the data?

I expect that for case A. I don't expect that for B or C.

File: scipy/stats/tests/test_morestats.py [hunk 5]
```diff
@@ -301,6 +305,71 @@ def test_gumbel_r(self):
...
+        assert res.statistic > res.critical_values[0]
+
+    def test_weibull_error(self):
+        # Check for helpful error message when Anderson test for Weibull fails
+        rng = np.random.default_rng(2044208699473992917)
+        c, loc, scale = 5, 0, 1
+        x = stats.weibull_min.rvs(c, loc, scale, size=100, random_state=rng)
+        message = "An error occured while fitting the Weibull distribution..."
+        with (pytest.raises(ValueError, match=message),
+              pytest.raises(FloatingPointError)):
```
Comment (mdhaber) [inline scipy/stats/tests/test_morestats.py:343]: I think this is the cause of the test failures. I never remember how to match multiple error types correctly in Python 3.8 (and before).

File: scipy/stats/tests/test_morestats.py [hunk 6]
```diff
@@ -15,10 +15,14 @@
...
 from scipy import stats
-from scipy.stats._morestats import _abw_state
+from scipy.stats._morestats import _abw_state, _get_As_weibull, _Avals_weibull
 from .common_tests import check_named_results
 from .._hypotests import _get_wilcoxon_distr, _get_wilcoxon_distr2
 from scipy.stats._binomtest import _binary_search_for_binom_tst
+from scipy.stats._distr_params import distcont
+
+
+distcont = dict(distcont)
```
Comment (mdhaber) [inline scipy/stats/tests/test_morestats.py:25]:
```suggestion
distcont = dict(distcont)  # type: ignore
```

**Review by mdhaber — COMMENTED** [review]
Should be pretty easy to fix test failures. I'll do that when gh-16848 and gh-16804 have merged. No sense in doing much else here until then.

**Comment by mdhaber** [comment]
I performed the following experiment with a lightly-modified version of f1a969664a5cf0425e2774c7fdd5bb3586793194 that causes the trial to fail when case B/C is detected, so this experiment is only for case A.

<details>

```python3
import numpy as np
from scipy import stats
from time import perf_counter
import matplotlib.pyplot as plt

rng = np.random.default_rng(1652117474769826144)
qrng = stats.qmc.Halton(5, seed=rng)

N = 10000
times = np.full(N, np.nan)
statistics = np.full(N, np.nan)
rejects = np.full((N, 8), np.nan)
params_true = np.full((N, 3), np.nan)
params_fit = np.full((N, 3), np.nan)


stop_at = None

for i in range(N):

    print(i)
    sample = qrng.random()
    sample = stats.qmc.scale(sample, [0, -3, -1, 2, -1], [2, 3, 2, 5, 1])[0]
    a, loc, scale, size = 10**sample[:4]
    loc_sign = np.sign(sample[4])
    loc *= loc_sign
    size = int(size)

    params_true[i] = [a, loc, scale]
    dist = stats.weibull_min(a, loc, scale)
    rvs = dist.rvs(size=size, random_state=rng)

    if stop_at is not None and i != stop_at:
        continue

    try:

        tic = perf_counter()
        res = stats.anderson(rvs, 'weibull_min')
        toc = perf_counter()
        times[i] = toc-tic
        statistics[i] = res.statistic
        rejects[i] = (res.statistic > res.critical_values)
        params_fit[i] = res.fit_result.params

    except Exception as e:
        print(f"Failure: {stats.weibull_min(rvs)} {e}")

    if i == stop_at:
        break

print(res.fit_result)
print(res.fit_result.params[0] - np.min(rvs))
res.fit_result.plot()

i_bad = np.isfinite(statistics)
good_rejects = rejects[i_bad]
good_statistics = statistics[i_bad]
good_times = times[i_bad]
good_reject_rate = np.sum(good_rejects, axis=0) / len(good_rejects)
```

</details>

$10,000$ sets of `weibull_min` parameters were quasi-randomly sampled log-uniformly in the ranges $c \in [1, 100]$ (shape), $\mu \in [10^{-3}, 10^{3}]$ (location, with equal chance of being positive or negative), and $\sigma \in [0.1, 100]$ (scale). For each set of parameters, a random sample of size quasi-randomly log-uniformly sampled on $n \in [10^2, 10^5]$ was drawn from the distribution. `anderson` was executed on the data, and I recorded whether the null hypothesis was rejected at each significance level (50% - 99.5%).

Out of the $10,000$ trials, $9,180$ were successful. Some trials failed because solving the likelihood equations was unsuccessful; others failed because case B or C was detected.

The nominal rates for rejection of the null hypothesis under the null distribution were:
array([0.5  , 0.25 , 0.15 , 0.1  , 0.05 , 0.025, 0.01 , 0.005])
and the observed rates were:
array([0.4866, 0.2363, 0.14  , 0.0929, 0.0435, 0.022 , 0.0105, 0.0053])

At most significance levels, the observed rates were a little lower than the nominal levels, indicating that the test is a little conservative at these levels (assuming that the user interprets the results as intended - only to gather evidence _against_ the null hypothesis). For the 99% and 99.5% significance levels, the null hypothesis was rejected a little more often than we would expect, but probably within the limits of random error, given the size of the experiment.

This shows promise. I think I would be happy releasing the test as-is, even if case B/C is not supported, as long as the documentation and error messages are explicit about the test's limitations. Of course, I'll work on improving it - just wanted to share preliminary results. I'd like to investigate:

- behavior with smaller samples $10 < n < 100$
- behavior with c < 1
- why solving the likelihood equations sometimes fails\
- behavior with data not drawn from the Weibull distribution (does it fail too often to make the test useful?)

**Comment by mdhaber** [comment]
@lispsil in my [comment above](https://github.com/scipy/scipy/pull/12787#issuecomment-1218410229) I explain that in numerical experiments, the critical values provided for small values of the Weibull shape parameter $m$ seem too low because they result in very high Type I error rates. The recommendation of the paper to use the critical values for $m=2$ for all smaller values of shape $m$ seems inappropriate, because $m << 2$ is qualitatively quite different from $m=2$ and tends to result in much higher values of Anderson $A^2$ statistic, even when the null hypothesis is true. 

What do you think about (for now, at least) raising an error when the fitted shape is less than $m$? 

We can save the case of $m < 2$ for another day. In the meantime, I think I should invest the time into providing a more complete, more general goodness-of-fit test using `monte_carlo_test` as outlined in items 3 and 4 [here](https://github.com/scipy/scipy/issues/10829#issuecomment-1179351989).

**Comment by lispsil** [comment]
@mdhaber definitely better to raise errors for small values of shape. Maybe we need to add the warning in the documentation that the one should be very cautious with `fit` for three parameters Weibull?

File: scipy/stats/tests/test_morestats.py [hunk 7]
```diff
@@ -310,9 +310,55 @@ def test_gumbel_r(self):
...
+        x = [225, 75, 57, 168, 107, 12, 61, 43, 29]
+        message = "Critical values of the test statistic are given for the..."
+        with pytest.warns(UserWarning, match=message):
+            stats.anderson(x, 'weibull_min')
+
     @pytest.mark.parametrize('distname',
                              ['norm', 'expon', 'gumbel_l', 'extreme1',
-                              'gumbel', 'gumbel_r', 'logistic'])
+                              'gumbel', 'gumbel_r', 'logistic', 'weibull_min'])
+
```
Comment (mdhaber) [inline scipy/stats/tests/test_morestats.py:361]:
```suggestion
```

File: scipy/stats/morestats.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/morestats.py]: We now use the analytical method of moments solution as a guess in `weibull_min.fit`

File: scipy/stats/morestats.py (hunk 1 above)
Comment (mdhaber) [reply scipy/stats/morestats.py]: Rather than round to an element of the dictionary, we round as suggested in the reference.

File: scipy/stats/tests/test_morestats.py (hunk 3 above)
Comment (mdhaber) [reply scipy/stats/tests/test_morestats.py]: Fitting is not foolproof, but it's reasonably reliable for Weibull-distributed data with moderate values of the parameters. 

image

I think that we only really need to distinguish between case A and case B/C in this PR. As I'll explain, we only really need to concern ourselves with case A.

For case C, the "likelihood steadily decreases as [the location] moves away from [the minimum of the data] towards $-\infty$", so the maximum likelihood occurs at the maximum possible value, which is constrained to be the minimum of the data. The authors also explicitly state that the minimum value of the data is the MLE of location in case B:
image
So case A is all cases in which the MLE of the location is not at the minimum of the data. 

I don't think the authors make it very clear, but case A tends to correspond with values of the shape parameter $m > 1$, and B/C is for $m < 1$. This makes sense from the plot of the PDF :
image

For small shape parameter, the PDF is highest toward the left end of the support, which is why it makes sense for `loc` to be at the minimum of the data.

This is important because the authors did not actually tabulate critical values for cases in which the shape parameter is less than 2, and I've observed that - even with their bias reduction procedure - the authors' recommendation to use the critical values for $m=2$ when $m < 2$ results in a very unconservative test. So I think whenever we find that the shape parameter $m < 2$, we should just report to the user that critical values are not available because they are below the tabulated range. We only need to provide a fit for and complete the test for case A; otherwise, we error out or report NaNs or something.

File: scipy/stats/morestats.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/morestats.py]: Done.

**Comment by mdhaber** [comment]
@chrisb83 I see that you started the review of this, so I thought you might be interested in taking another look. So that you don't have to dig through all the comments above, here is a summary.

The goal of this PR was to perform the unknown-parameter Anderson-Darling test for the three-parameter Weibull distribution using the critical values provided by [Lockhart and Stephens](https://www.jstor.org/stable/2346122?seq=1). (There may be more recent papers that offer a greater range of critical values, and that would be a useful enhancement, but I've limited the scope of my work to the table given by Lockhart and Stephens.)

The paper recommends that the critical values are accurate for sample size $n\geq10$, but they do depend on shape parameter $m$. The paper gives critical values for $0 < c = 1/m < 0.5$ ($2 < m < \infty$) and states that if $c \geq 0.5$, the values corresponding with $c \geq 0.5$ should be used.

In the "Case A" of the paper, fitting is relatively straightforward. One of the likelihood equations is solved in closed form, and the other two are solved numerically. I start with `weibull_min.fit` and refine the solution by solving the likelihood equations (5-7 in the paper).

The other cases, B and C, are the tricky ones and lead to most of the complexity of the paper. But I've found that these cases correspond with $c \geq 1$, which is beyond the range of tabulated values, and although the paper recommends using the values for $c = 0.5$ in this case, I've found that this does not control the false positive rate of the test. Assuming the user is using the test only as evidence _against_ the null hypothesis, this is not conservative. So I have actually removed the special code I wrote that corresponds with cases B and C, and now I simply abort the test if fitting converges to a shape parameter less than 2 or if case B/C is detected based on the maximum likelihood estimate of the location being at the minimum of the data. (I may reduce the shape parameter threshold depending on further testing.)

If you agree that it is still useful to add a test that fails sometimes and, in this case, raise an informative error message, review should be pretty straightforward. At this point, I would prefer to go ahead and merge this simple version of the test, despite the fact that it will not work for all data, rather than leave it open forever.

When this is merged, I'll open a PR for a `goodness_of_fit` test using `monte_carlo_test`. Rather than just critical values and significance levels, this would provide accurate $p$-values regardless of sample size/parameter values for any distribution which can be fit reliably(and in reasonable time). The user will be able to choose the desired test statistic (e.g. KS, Anderson-Darling... someday Shapiro-Wilks), select the fitting procedure, and specify known parameters.

**Comment by mdhaber** [comment]
> Maybe we need to add the warning in the documentation that the one should be very cautious with fit for three parameters Weibull?

I think the fits in these cases is OK. The problem is mostly the critical values. In any case, I'm raising an error when either the shape is too in small or the location equals the minimum of the data. They almost always coincide. 

This is basically in final form, from my perspective. If it looks good to you, would you review and click the approve button when you submit your review?

File: scipy/stats/_morestats.py [hunk 8]
```diff
@@ -1830,6 +1831,91 @@ def shapiro(x):
...
+                  "performing a custom goodness-of-fit test using "
+                  "`scipy.stats.monte_carlo_test`.")
+
+    if np.allclose(u, np.min(x)) or m < 1:
+        # The critical values provided by [7] don't seem to control the
+        # Type I error rate in this case. Error out.
+        message = ("The maximum likelihood estimate of the has converged to "
+                   "a solution in which the location is equal to the minimum "
+                   "of the data, the shape parameter is less than 1, or both. "
+                   "The critical values used by `anderson` (see [7]) are "
```
Comment (chrisb83) [inline scipy/stats/_morestats.py:1896]: "are known" sounds like this is a well-known fact. is that true or is it your testing that indicates this?

File: scipy/stats/_morestats.py [hunk 9]
```diff
@@ -1969,6 +2074,25 @@ def rootfunc(ab, xj, N):
...
+        # [7] writes our 'c' as 'm', and they write `c = 1/m`. Use their names.
+        m, loc, scale = distributions.weibull_min.fit(y)
+        m, loc, scale = _weibull_fit_check((m, loc, scale), y)
+        fit_params = m, loc, scale
+        logcdf = stats.weibull_min(*fit_params).logcdf(y)
+        logsf = stats.weibull_min(*fit_params).logsf(y)
+        c = 1 / m  # m and c are as used in [7]
+        sig = array([0.5, 0.75, 0.85, 0.9, 0.95, 0.975, 0.99, 0.995])
+        critical = _get_As_weibull(c)
+        # Goodness-of-fit tests should only be used to provide evidence
```
Comment (chrisb83) [inline scipy/stats/_morestats.py:2093]: I did not understand this comment. if we adjust the values from [7], wouldn't this need to be explained in the docstring?

**Review by chrisb83 — COMMENTED** [review]
I only have a few comments, the PR looks good overall

File: scipy/stats/_morestats.py (hunk 8 above)
Comment (mdhaber) [reply scipy/stats/_morestats.py:1896]: Good call. My testing. Should I make that explicit?

File: scipy/stats/_morestats.py (hunk 9 above)
Comment (mdhaber) [reply scipy/stats/_morestats.py:2093]: I'm not sure if this is significant. The values in [7] are only provided to three decimal places, and they recommend interpolating. I don't think we should return the full precision after interpolation because most of the figures are not significant. This just rounds to three decimal places like the tabulated values, and the comment explains why I'm taking care to round _up_.

File: scipy/stats/tests/test_morestats.py [hunk 10]
```diff
@@ -310,9 +310,47 @@ def test_gumbel_r(self):
...
+        As_ref = As40 + (c - 0.4)/(0.45 - 0.4) * (As45 - As40)
+        # atol=1e-3 because results are rounded up to the next third decimal
+        assert np.all(res.critical_values > As_ref)
+        assert_allclose(res.critical_values, As_ref, atol=1e-3)
+
+    def test_weibull_min_case_B(self):
+        # From `anderson` reference [7]
+        x = np.array([74, 57, 48, 29, 502, 12, 70, 21,
+                      29, 386, 59, 27, 153, 26, 326])
+        message = "The maximum likelihood estimate of the has converged to "
```
Comment (mdhaber) [inline scipy/stats/tests/test_morestats.py:336]:
```suggestion
        message = "Maximum likelihood estimation has converged to "
```

File: scipy/stats/_morestats.py [hunk 11]
```diff
@@ -1830,6 +1831,91 @@ def shapiro(x):
...
+
+    suggestion = ("Maximum likelihood estimation is known to be challenging "
+                  "for the three-parameter Weibull distribution. Consider "
+                  "performing a custom goodness-of-fit test using "
+                  "`scipy.stats.monte_carlo_test`.")
+
+    if np.allclose(u, np.min(x)) or m < 1:
+        # The critical values provided by [7] don't seem to control the
+        # Type I error rate in this case. Error out.
+        message = ("The maximum likelihood estimate of the has converged to "
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_morestats.py:1893]: Oops! How about:
```suggestion
        message = ("Maximum likelihood estimation has converged to "
```

File: scipy/stats/_morestats.py [hunk 12]
```diff
@@ -1830,6 +1831,91 @@ def shapiro(x):
...
+                   "a solution in which the location is equal to the minimum "
+                   "of the data, the shape parameter is less than 1, or both. "
+                   "The critical values used by `anderson` (see [7]) are "
+                   "known to be too low in this case. " + suggestion)
```
Comment (mdhaber) [inline scipy/stats/_morestats.py:1897]: How about I state something that is definitely true:
```suggestion
                   "of the data, the shape parameter is less than 2, or both. "
                   "The table of critical values in [7] does not "
                   "include this case. " + suggestion)
```
since the table only does not have rows for $c = 1/m > 0.5$?

File: scipy/stats/_morestats.py [hunk 13]
```diff
@@ -1969,6 +2074,25 @@ def rootfunc(ab, xj, N):
...
+        m, loc, scale = _weibull_fit_check((m, loc, scale), y)
+        fit_params = m, loc, scale
+        logcdf = stats.weibull_min(*fit_params).logcdf(y)
+        logsf = stats.weibull_min(*fit_params).logsf(y)
+        c = 1 / m  # m and c are as used in [7]
+        sig = array([0.5, 0.75, 0.85, 0.9, 0.95, 0.975, 0.99, 0.995])
+        critical = _get_As_weibull(c)
+        # Goodness-of-fit tests should only be used to provide evidence
+        # _against_ the null hypothesis. Be conservative and round up.
+        critical = np.round(critical + 0.0004999999, decimals=3)
```
Comment (mdhaber) [inline scipy/stats/_morestats.py:2095]: Here's something I can do to make this a little less mysterious.
```suggestion
        critical = np.round(critical + 0.0005, decimals=3)
```
I thought that e.g. `1.000 + 0.0005 ` would round up to `1.001`, but it does not. But `1.0001 + 0.0005` will round up to `1.001`, which was my intent.

**Review by mdhaber — COMMENTED** [review]
Thanks for the review @chrisb83!
I think I replied to your comments. Please commit the ones that you agree with and let me know if there are other comments.

File: scipy/stats/_morestats.py (hunk 8 above)
Comment (chrisb83) [reply scipy/stats/_morestats.py:1896]: yes, I think it would be good to rephrase in that case

Comment (chrisb83) [reply scipy/stats/_morestats.py:1896]: oh, you already did... i will apply your suggestion

File: scipy/stats/_morestats.py (hunk 9 above)
Comment (chrisb83) [reply scipy/stats/_morestats.py:2093]: ok, that makes sense

**Comment by mdhaber** [comment]
@chrisb83 were there other comments for this one?

**Comment by mdhaber** [comment]
@chrisb83 It sounded like this looked close to you before, so I thought I'd check in again. Happy New Year!

**Comment by chrisb83** [comment]
Hi Matt, I think this looks good, there is a clear focus on case A including useful error messages if case B/C are detected. The test cases look good + your analysis above for the critical values. I will approve the PR now.

**Review by chrisb83 — APPROVED** [review]
see previous comment

**Comment by chrisb83** [comment]
@mdhaber I was about to merge this PR when I noted the test failures. it looks unrelated, all three tests fail when trying to clone boost. do you know why?

**Comment by mdhaber** [comment]
No, but it looks an unrelated server timeout.

**Comment by mdhaber** [comment]
I think it's safe to ignore.

**Comment by mdhaber** [comment]
Thanks for the review @chrisb83!
